### DISCUSSION TITLE
Support 4-digit TM1637-based displays

### DISCUSSION THREAD

**Discussion by neilenns** [MobiFlight/MobiFlight-Connector#604] [Ideas]
**Is your feature request related to a problem? Please describe.**
Support displays like this: https://www.amazon.com/s?k=TM1637+display&ref=nb_sb_noss_2

**Describe the solution you'd like**
Arduino library: https://github.com/avishorp/TM1637

**Describe alternatives you've considered**
MAX7219s

**Additional context**
This has come up twice in the last couple of weeks in Discord.

**Comment by tigert** [comment]
https://lygte-info.dk/project/DisplayDriver%20UK.html

Found this, no idea if this is good, huge, or something useful, but might be worth a look?

_"I needed a library for a 7 segment display, but was not satisfied with the Arduino libraries, as a result I have made a new "library" that can handle nearly every cheap display on Ebay (See above picture, they are all supported) and has functions to display int/long/float/double/hex/binary/text. The library is fast and fairly compact._

_The displays supported include: HC164, HC595, HT1621, HT16K33, MAX7219, TM1637 and TM1638"_

**Comment by tigert** [comment]
and some critique on the above lib: https://forum.arduino.cc/t/library-for-many-7-14-16-segment-serial-displays/594987/9

**Comment by tigert** [comment]
it seems to use lots of defines that you choose at compile time to select what chip you have, which might be a problem for a generic use like our case is.

**Comment by MobiFlight-Admin** [comment]
Max7219 are darn cheap and available to everyone. What are the benefits of TM1637?

**Reply by neilenns** [reply]
Apparently they're commonly used with other flight sim integration systems. I opened this when we had a run of 2-3 people in the span of a week asking about it in Discord.

**Reply by tigert** [reply]
There are no 4 digit MAX modules - I guess people might mistakenly buy those 4 digit TM1637 modules since they look very similar to the MAX ones. At least I got a few by accident without realizing they have a different chip.

**Reply by CsabaKafka** [reply]
Personaly I would use them on an MCP panel for IAS for example. I think lot of people would like to use them, especially when they build a B737 cockpit. I have couple of them and can´t really use them.

**Reply by Drboratti** [reply]
> Max7219 are darn cheap and available to everyone. What are the benefits of TM1637?

Y es... You are right.
But there are a lot of people that have been making their cockpits for years to use with xplane... In that way, simvim was the gold standard for that moment.
Now MS Flight Simulator is back, and a lot of virtual pilots that were using xplane are migrating to the New sim. I wouldn't be fair nor good for them to Start all the home cockpit from zero.

Mobiflight has the opotunity of being the New gold standard for all sims, but need to support several New things like tm1637
Besides... Max7219 has some drawbacks. If you get the driver alone, sometimes it gets really hard to find 3 or four dígits display s un tha same size. At least Where I live.

Btw... Thanks for MF... Its an outstanding job

**Reply by Drboratti** [reply]
Making an b737 cockpit with all 8 dígits would be a pain

**Reply by DocMoebiuz** [reply]
the standard 8 digit max modules with the detachable digits can easily split up in two groups of four digits each.

we also have pcb adapters for combinations of 3 to 6 digits.

I don't know what the footprint of driving tm1637 would  be in the firmware and we are experiencing limitations already. So its not about whether to add it or not but also if it really still fits into the firmware.

**Reply by Drboratti** [reply]
> the standard 8 digit max modules with the detachable digits can easily split up in two groups of four digits each.
> 
> we also have pcb adapters for combinations of 3 to 6 digits.
> 
> I don't know what the footprint of driving tm1637 would be in the firmware and we are experiencing limitations already. So its not about whether to add it or not but also if it really still fits into the firmware.

Well... that's a new hole story. Now I understand, and of course there are many more important things to take care of, so let's save some space in the firmware. But... there is still something I don't understand: Simvim works with a single arduino mega 2560, and supports hardware that MF doesn't support, such as TM1637, IR positioning sensor (TCRT5000), encoders in MUX. And the firmware must be the same size. I should know... I'm just a user. I don't have any programming skills... what's more, I may have a low IQ. So... I'm not a qualified person to talk about it. And my intention is only to help... so if there is any way I could help, count on me. And thank you for the good job.

**Reply by jeefo93** [reply]
I already have a fully assembled 737/MCP on TM1637 built for X-Plane in conjunction with SimVim (HCSCIS). I chose the TM1637 due to the ready size modules available - 4-bit for CRS/HDG/CRS and 6-bit for everything else. Installed - connected pins - assigned pins - it works.
To upgrade to the Max7219, my MCP must be completely rebuilt. In addition, DM13A drivers are used to connect the LED indicators of MCP buttons, FD and A/T, which is also not supported by MobiFlight.

There are some other developments like libraries for Arduino using the SimConnect API. Theoretically, with their help, you can write suitable code for Arduino. However, in the case of PMDG's 737-700, they are useless, since by default SimConnect does not provide access to the L:Vars variables in which the simulator stores MCP display values.

I am delighted with the work you have done to create this product. But, unfortunately, even though the code is in the public domain, I can’t take and rebuild the build for my needs, because I don’t have enough skills for this. Therefore, it remains only to fly MSFS2020 without a hardware MCP and hope that such electronic components are used not only by me, and someday you will add their support to your product.


![2022-06-20 18 00 00_cr](https://user-images.githubusercontent.com/108031398/175974637-503388b7-6248-42f1-9575-7b912787d1c8.jpg)
![2022-06-20 08 27 28 HEIC](https://user-images.githubusercontent.com/108031398/175975181-11457a8b-d0a7-4e99-b484-150f8c8fb924.jpeg)

**Comment by tigert** [comment]
there is now a pull request for this: #1252
